src/Grid/grid.jl

```diff
@@ -707,12 +707,91 @@ function _addset!(grid::AbstractGrid, name::String, f::Function, _ftype::Functio
     dict[name] = _set
     grid
 end
+"""
+    getfaceedges(face::NTuple{n,Int64})
+
+Returns the edges in a given face.
+
+```julia
+getfaceedges(face)
+```
+"""
+function getfaceedges(face::NTuple{n,Int64}) where {n}
+    ntuple(i -> i!=n ? (face[i+1],face[i]) : (face[1],face[n]),n)
+end
+"""
+    getfacevertices(face::NTuple{n,Int64})
+
+Returns the vertices in a given face.
+
+```julia
+getfacevertices(face)
+```
+"""
+function getfacevertices(face::NTuple{n,Int64}) where {n}
+    face
+end
+"""
+    getedgevertices(edge::NTuple{2,Int64})
+
+Returns the vertices in a given edge.
+
+```julia
+getedgevertices(edge)
+```
+"""
+function getedgevertices(edge::NTuple{2,Int64})
+    edge
+end
```

Comment:
Sorry I was not very specific here. Please note that the sets in the grid uses `EdgeIndex`, `FaceIndex` and `VertexIndex` for the sets (see https://github.com/Ferrite-FEM/Ferrite.jl/blob/master/src/Grid/grid.jl#L349-L351), so one way to add edges and vertices via the `ExclusiveTopology` would be to get the corresponding `EdgeIndex` and/or `VertexIndex` of a boundary face. You can move this code below into your functions as anonymous functions.

Reply:
Thanks for the comments. Do you mean the helpers should return `EdgeIndex`/`VertexIndex` given `FaceIndex`/`EdgeIndex` as Input? Maybe something like this:
```Julia
function _getentities(grid::AbstractGrid, parent::BoundaryIndex, f::Function, _ftype::Function, _fentities::Function, _fparent::Function, _indextype::Type; all::Bool=true)
    _set = Set{_indextype}()
    cell = getcells(grid)[parent[1]]
    cell_entities = _ftype(cell)
    entities = _fentities(_fparent(cell)[parent[2]])#_fparent returns global vertices ids of face/edge. _fentities transforms them to pairs for edges.
    for entity_idx in findall(x->x ∈ entities,cell_entities) # entities can be edges/vertices in the face/edge
        pass = all
            for node_idx in cell_entities[entity_idx] # using cell entities common with boundary face
                v = f(grid.nodes[node_idx].x)
                all ? (!v && (pass = false; break)) : (v && (pass = true; break))
            end
        pass && push!(_set, _indextype(parent[1], entity_idx))
    end
    return _set
end
```
Which is then called from `getfaceedges`, `getfacevertices`, `getedgevertices` Like this:
```Julia
getfaceedges(grid::AbstractGrid, parent::BoundaryIndex, f::Function; all::Bool=true) =
    _getentities(grid, parent, f, Ferrite.edges,
     face -> ntuple(i -> length(face)>2&&i!=length(face) ? (face[i+1],face[i]) : (face[1],face[length(face)]),length(face)), 
     Ferrite.faces, EdgeIndex; all = all)
```
```Julia
getfacevertices(grid::AbstractGrid, parent::BoundaryIndex, f::Function; all::Bool=true) =
    _getentities(grid, parent, f, Ferrite.vertices, face -> face,Ferrite.faces, VertexIndex; all = all)
```
```Julia
getedgevertices(grid::AbstractGrid, parent::BoundaryIndex, f::Function; all::Bool=true) =
    _getentities(grid, parent, f, Ferrite.vertices, edge -> edge,Ferrite.edges, EdgeIndex; all = all)
```
with 
```Julia
#Just for validating that the function f applies to the face nodes
_checkface(grid::AbstractGrid, parent::BoundaryIndex, f::Function; all::Bool=true) = 
    _getentities(grid, parent, f, Ferrite.faces, face -> (face,),Ferrite.faces, FaceIndex; all = all)
``` 
for testing faces against function `f`
And used in `addboundaryset` like this:
```Julia
addboundaryfaceset!(grid::AbstractGrid, topology::ExclusiveTopology, name::String, f::Function; all::Bool=true) =
    _addset!(grid, topology, name, f, grid.facesets, _checkface, FaceIndex; all=all)
addboundaryedgeset!(grid::AbstractGrid, topology::ExclusiveTopology, name::String, f::Function; all::Bool=true) =
    _addset!(grid, topology, name, f, grid.edgesets, getfaceedges, EdgeIndex; all=all)
addboundaryvertexset!(grid::AbstractGrid, topology::ExclusiveTopology, name::String, f::Function; all::Bool=true) =
    _addset!(grid, topology, name, f, grid.vertexsets, getfacevertices, VertexIndex; all=all)
function _addset!(grid::AbstractGrid, topology::ExclusiveTopology, name::String, f::Function, dict::Dict, _fentities::Function, _indextype::Type; all::Bool=true)
    _check_setname(dict, name)
    _set = Set{_indextype}()
    for face in findall(isempty,topology.face_neighbor)
        union!(_set,_fentities(grid, _indextype(face[1],face[2]), f; all = all))
    end
    _warn_emptyset(_set, name)
    dict[name] = _set
    grid
end
```
Is this what you mean?
I tested it with the `Triangle` and `Tetrahedron` meshes. More details on that is the testing comment.
Thanks!

Reply:
This is indeed what I had in mind, thanks! I thought maybe this is an easier API which could be internally utilized beyond this PR. :)

Reply:
I think the new methods should get `FaceIndex`/`EdgeIndex` respectively instead of `BoundaryIndex` and  `union!(_set,_fentities(grid, _indextype(face[1],face[2]), f; all = all))` should not take `_indextype` as input, but a new type `_boundarytype` or something along these lines. I will try to take some time to investigate this later this week.

Reply:
Sorry, turns out extracting all boundary edges in local definitions (i.e. as `EdgeIndex`) is a bit more complicated, because the test is not local. Your code seems to work as expected after some quick local tests on my machine.
To elaborate on the problem: Elements without faces on the boundary still can have boundary edges, see https://cecas.clemson.edu/cvel/modeling/EMAG/EMAP/emap4/Meshing.html Fig. 3 for an example. The same problem appears for vertices. 

This problem should not affect Ferrite internals when applying boundary condition, but may surprise users which need the explicit representation for all elements (e.g. when doing some fancy integration of edges). Maybe @lijas can comment on this?

Reply:
You're right, I changed the types in the methods, thanks!
 I think we can extract the boundary faces then extract the edges from all cells that match those from the boundary face like this:
```Julia
function _getentities(grid::AbstractGrid, parent::BoundaryIndex, f::Function, _ftype::Function, _fentities::Function, _fparent::Function, _indextype::Type; all::Bool=true)
    _set = Set{_indextype}()
    cells = getcells(grid)
    parent_cell = cells[parent[1]]
    entities = _fentities(_fparent(parent_cell)[parent[2]])#_fparent returns global vertices ids of face/edge. _fentities transforms them to pairs for edges.
    for (cell_idx, cell) in enumerate(cells)
        cell_entities = _ftype(cell)
        for entity_idx in findall(x->x ∈ entities, cell_entities) # entities can be edges/vertices in the face/edge
            pass = all
                for node_idx in cell_entities[entity_idx] # using cell entities common with boundary face
                    v = f(grid.nodes[node_idx].x)
                    all ? (!v && (pass = false; break)) : (v && (pass = true; break))
                end
            pass && push!(_set, _indextype(cell_idx, entity_idx))
        end
    end
    return _set
end
```
This should return all the indices of the edges/nodes at the boundary even if the cell does not have boundary faces. the downside of this is that we're iterating over all cells, yet I think there are better implementations (maybe extract those of the boundary faces first then only search the grid if vertices/nodes are needed). This way `getfaceedges` will return edges that do not belong to the cell/face but are defined using the same nodes. Please let me know what you think.
I tested this with the test in the last reply on the test section and it passed the test for edges, faces and vertices.
> ```julia
> # Triangle grid test
>     grid = generate_grid(Triangle,(2,2))
>     topology = ExclusiveTopology(grid)
>     @test extractboundary(grid,topology) == extractboundarycheck(grid)
> # Tetrahedron grid test
>     grid = generate_grid(Tetrahedron,(10,10,10))
>     topology = ExclusiveTopology(grid)
>     @test extractboundary(grid,topology) == extractboundarycheck(grid)
> ```
Thanks!